packages/theme/src/cli/utilities/theme-environment/hot-reload/server.ts

```diff
@@ -0,0 +1,232 @@
+import {getClientScripts, HotReloadEvent} from './client.js'
+import {render} from '../storefront-renderer.js'
+import {patchRenderingResponse} from '../proxy.js'
+import {createEventStream, defineEventHandler, getProxyRequestHeaders, getQuery, sendError, type H3Error} from 'h3'
+import {renderWarning} from '@shopify/cli-kit/node/ui'
+import {extname} from '@shopify/cli-kit/node/path'
+import {parseJSON} from '@shopify/theme-check-node'
+import EventEmitter from 'node:events'
+import type {Theme, ThemeFSEventPayload} from '@shopify/cli-kit/node/themes/types'
+import type {DevServerContext} from '../types.js'
+
+// --- Template Replacers ---
+
+/** Store which files are currently only updated in-memory, not in remote */
+const inMemoryTemplateFiles = new Set<string>()
```

Comment:
Exactly, that captures the item 5 in the list below. Our final state with the `inMemoryTemplateFiles` is:

1. The CLI stores everything in the `inMemoryTemplateFiles`
2. The CLI displays links so users can preview their theme in the browser
3. The CLI uploads files and removes them from `inMemoryTemplateFiles` as each file is uploaded, helping to avoid performance impact on the server
4. The `inMemoryTemplateFiles` becomes empty when uploads are finished
5. The `inMemoryTemplateFiles` is updated when users apply changes to files

Currently, the CLI is blocking users and uploading everything in advance, but we may iterate on this when we get the `localThemeFileSystem.addEventListener` implemented 🚀

Reply:
Absolutely! I missed capturing that in the message above. Certainly, some elements of the theme block the first rendering and should be uploaded in advance. I've updated the message to reflect that 🚀 Thank you, @frandiox!